src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -178,4 +193,199 @@ module Ast =
+                acc @ Error [ sprintf "Field '%s' of '%s' type is of type kind %s, and therefore should not contain inner fields in its selection." selection.Name selection.ParentType.Name (fieldType.Kind.ToString()) ]
+            | TypeKind.INTERFACE | TypeKind.UNION | TypeKind.OBJECT when selectionSetLength = 0 ->
+                acc @ Error [ sprintf "Field '%s' of '%s' type is of type kind %s, and therefore should have inner fields in its selection." selection.Name selection.ParentType.Name (fieldType.Kind.ToString()) ]
+            | _ -> acc
+        let acc =
+            match selection.FieldType with
+            | Some fieldType -> validateByKind acc fieldType selection.SelectionSet.Length
+            | None -> acc
+        selection.SelectionSet |> List.fold (fun acc selection -> checkLeafFieldSelection acc selection) acc
+
+    let validateLeafFieldSelections (schemaInfo : SchemaInfo) (ast : Document) =
+        let fragmentDefinitions = getFragmentDefinitions ast
+        ast.Definitions
+        |> List.choose (tryGetSelectionSetAndType schemaInfo)
+        |> List.fold (fun acc (objectType, selectionSet) ->
+            let set = getSelectionSetInfo schemaInfo fragmentDefinitions objectType selectionSet
+            set |> List.fold (fun acc selection -> checkLeafFieldSelection acc selection) acc) Success
+
+    let rec private checkFieldArgumentNames (acc : ValidationResult) (schemaInfo : SchemaInfo) (selection : SelectionInfo) =
+        match selection.FieldType with
+        | Some fieldTypeRef ->
+            let acc =
+                selection.Arguments
+                |> List.fold (fun acc arg ->
+                    match selection.ArgDefs |> Option.map (Array.tryFind (fun d -> d.Name = arg.Name)) |> Option.flatten with
+                    | Some _ -> acc
+                    | None -> acc @ Error [ sprintf "Field '%s' of type '%s' does not have an input named '%s' in its definition." selection.Name selection.ParentType.Name arg.Name ]) acc
+            selection.Directives
+            |> List.fold (fun acc directive ->
+                match schemaInfo.Directives |> Array.tryFind (fun d -> d.Name = directive.Name) with
+                | Some directiveType ->
+                    directive.Arguments
+                    |> List.fold (fun acc arg ->
+                        match directiveType.Args |> Array.tryFind (fun argt -> argt.Name = arg.Name) with
+                        | Some _ -> acc
+                        | None -> acc @ Error [ sprintf "Directive '%s' of field '%s' of type '%s' does not have an argument named '%s' in its definition." directiveType.Name selection.Name selection.ParentType.Name arg.Name ]) acc
+                | None -> acc) acc
+        | None -> acc
+
+    let validateArgumentNames (schemaInfo : SchemaInfo) (ast : Document) =
+        let fragmentDefinitions = getFragmentDefinitions ast
+        ast.Definitions
+        |> List.choose (tryGetSelectionSetAndType schemaInfo)
```

Comment:
Could this pattern be factored out into a new function?

Reply:
The code did change on this part, and now is using only `List.fold`.